rep-0008.rst

```diff
@@ -22,27 +25,90 @@ Please follow Python's PEP 8 [1]_ for guidelines on formatting, naming, etc...
+The recommended layout is to put all of your package's importable Python code in the ``src`` folder of your package and in a "Python package" with the same name as your package.
+In this case a "Python package" is a folder which contains a ``__init__.py`` file, indicating that the folder contains additional Python packages and modules.
+
+Here is the recommended layout of an example ``catkin`` package called ``my_package`` which contains Python code::
+
+    my_package
+    ├── package.xml
+    ├── setup.py
+    └── src
+        └── my_package
+            ├── __init__.py
+            ├── a_nested_python_package
+            │   └── __init__.py
+            └── a_python_module.py
+
+In this example there is a root Python package in the ``src`` folder called ``my_package``, which follows the recommendation.
+This root Python package then contains a Python module called ``a_python_module`` and a nested Python package called ``a_nested_python_package``.
+
+In this example the ``setup.py`` will use the convenience function ``generate_distutils_setup``, which is provided by ``catkin``, in order to extract information like the version and description from the
+package's manifest file, i.e. the ``package.xml``:
+
+.. code-block:: python
+
+    from distutils.core import setup
+    from catkin_pkg.python_setup import generate_distutils_setup
+
+    d = generate_distutils_setup(
+        packages=['my_package', 'my_package.a_nested_python_package'],
+        package_dir={'': 'src'},
+    )
+
+    setup(**d)
+
+Note that the ``packages`` parameter points to all Python packages which should be installed, even the nested one.
+Also note the ``package_dir`` parameter indicates that all aforementioned packages are located in the ``src`` folder by using the empty string as a key.
+This standard boiler plate configuration for Python packages allows you to place your Python code anywhere you like within your package and follows the Python ``setuptools`` conventions.
+
+If your package contains messages or services then they will generate code into a Python package called ``msgs`` or ``srvs`` in the root Python package, e.g. ``my_package/src/my_package/msgs`` or ``my_package/src/my_package/srvs``.
+Though this generated code does not go into your package's source code, it does get merged with your package's Python source when getting installed.
+This is a new mechanism introduced with ``catkin`` because ``catkin`` installs into a FHS [5]_ layout, whereas ``rosbuild`` did not have an install step.
+
+Your package shares a global ``PYTHONPATH`` namespace with other packages, so be considerate when creating Python packages at the root of your package's ``src`` folder as they can collide with others.
+
+Any executable file, including Python nodes and scripts, are accessible through ``rosrun``.
```

Comment:
This holds for ones installed by catkin. We should explicitly discourage installing via setup.py as it will install outside the package scope.